algorithmic_efficiency/workloads/librispeech_deepspeech/librispeech_jax/models.py

```diff
@@ -0,0 +1,730 @@
+      nn.transforms.scan,
+      variable_broadcast='params',
+      in_axes=(1, flax.core.axes_scan.broadcast, flax.core.axes_scan.broadcast),
+      out_axes=1,
+      split_rngs={'params': False})
+  def unroll_cell(self,
+                  cell_state: StateType,
+                  inputs: Array,
+                  recurrent_dropout_mask: Optional[Array],
+                  deterministic: bool):
+    """Unrolls a recurrent cell over an input sequence.
+
+    Args:
+      cell_state: The initial cell state, shape: <float32>[batch_size,
+        hidden_size] (or an n-tuple thereof).
+      inputs: The input sequence. <float32>[batch_size, seq_len, input_dim].
+      recurrent_dropout_mask: An optional recurrent dropout mask to apply in
+        between time steps. <float32>[batch_size, hidden_size].
+      deterministic: Disables recurrent dropout when set to True.
+
+    Returns:
+      The cell state after processing the complete sequence (including padding),
+      and a tuple with all intermediate cell states and cell outputs.
+    """
+    # We do not directly scan the cell itself, since it only returns the output.
+    # This returns both the state and the output, so we can slice out the
+    # correct final states later.
+    new_cell_state, output = self.cell(cell_state, inputs)
+    return new_cell_state, (new_cell_state, output)
+
+  def __call__(self,
+               inputs: Array,
+               lengths: Array,
+               initial_state: StateType,
+               reverse: bool = False,
+               deterministic: bool = False):
+    """Unrolls the RNN cell over the inputs.
+
+    Arguments:
+      inputs: A batch of sequences. Shape: <float32>[batch_size, seq_len,
+        input_dim].
+      lengths: The lengths of the input sequences.
+      initial_state: The initial state for the RNN cell. Shape: [batch_size,
+        hidden_size].
+      reverse: Process the inputs in reverse order, and reverse the outputs.
+        This means that the outputs still correspond to the order of the inputs,
+        but their contexts come from the right, and not from the left.
+      deterministic: Disables recurrent dropout if set to True.
+
+    Returns:
+      The encoded sequence of inputs, shaped <float32>[batch_size, seq_len,
+        hidden_size], as well as the final hidden states of the RNN cell. For an
+        LSTM cell the final states are a tuple (c, h), each shaped <float32>[
+          batch_size, hidden_size].
+    """
+    if reverse:
+      inputs = flip_sequences(inputs, lengths)
+
+    recurrent_dropout_mask = None
+    _, (cell_states, outputs) = self.unroll_cell(initial_state,
+                                                 inputs,
+                                                 recurrent_dropout_mask,
+                                                 deterministic)
+    final_state = jax.tree_map(
+        lambda x: x[jnp.arange(inputs.shape[0]), lengths - 1], cell_states)
```

Comment:
as far as I can tell, we never use this, so lets delete this code

Reply:
deleted